[MemCpyOpt] Drop dead `memmove` calls on `memset`'d source data

antoniofrighetto opened the pull request:
When a memmove happens to clobber source data, and such data have been previously memset'd, the memmove may be redundant.

Fixes: https://github.com/llvm/llvm-project/issues/101680.

DianQK commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1854,12 +1855,53 @@ bool MemCpyOptPass::processMemCpy(MemCpyInst *M, BasicBlock::iterator &BBI) {
 
+/// Memmove calls with overlapping src/dest buffers that come after a memset may
+/// be removed.
+bool MemCpyOptPass::mayBeMemMoveMemSetDependency(MemMoveInst *M) {
```
I guess this should be "must"?

antoniofrighetto replied:
I think it's implicit in the fact that the memmove comes after the memset?

aeubanks replied:
`may` in the function name is not strong enough, it should be something like `isMemMove...`

DianQK commented on llvm/test/Transforms/MemCpyOpt/fca2memcpy.ll:
```diff
@@ -51,8 +51,8 @@ define void @destroysrc(ptr %src, ptr %dst) {
-; CHECK-NEXT:    call void @llvm.memcpy.p0.p0.i64(ptr align 8 [[DST:%.*]], ptr align 8 [[SRC]], i64 16, i1 false)
-; CHECK-NEXT:    call void @llvm.memset.p0.i64(ptr align 8 [[SRC:%.*]], i8 0, i64 16, i1 false)
+; CHECK-NEXT:    call void @llvm.memcpy.p0.p0.i64(ptr align 8 [[DST:%.*]], ptr align 8 [[SRC:%.*]], i64 16, i1 false)
+; CHECK-NEXT:    call void @llvm.memset.p0.i64(ptr align 8 [[SRC]], i8 0, i64 16, i1 false)
```
It looks like a NFC change?

antoniofrighetto replied:
Right, it can be landed as a separate change (or I can drop it), if we wish so.

dianqk reviewed (commented):
Is there any real-world code written like this?

antoniofrighetto commented:
> Is there any real-world code written like this?

While the usefulness of the patch is indeed questionable, dubious at the very least, I thought it could still make sense considering that the underlying optimization is quite trivial/basic per sé, and has a low compile-time. I feel like compilers should still perform at their best if possible, while clearly trading for efficiency. That said, I advocate for optimizations that turn out to be actually profitable, so I do not oppose if this is an unnecessary addition.

antoniofrighetto commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1854,12 +1855,53 @@ bool MemCpyOptPass::processMemCpy(MemCpyInst *M, BasicBlock::iterator &BBI) {
+    // On the off-chance the memmove clobbers src with previously memset'd
+    // bytes, the memmove may be redundant.
+    if (mayBeMemMoveMemSetDependency(M)) {
+      LLVM_DEBUG(dbgs() << "Removed redundant memmove.\n");
```
The logging seems to be already coherent with the ones used in MemCpyOpt? E.g.: https://github.com/llvm/llvm-project/blob/71051deff27928cff908ea794e09806eee662801/llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp#L1826

DianQK replied:
Ah, I like `LLVM_DEBUG(dbgs() << "MemCpyOptPass: Optimizing memmove -> memcpy: " << *M << "\n");`, but it doesn't seem to be a problem.

dtcxzyw commented:
> Is there any real-world code written like this?

See https://github.com/dtcxzyw/llvm-opt-benchmark/pull/1135/files

antoniofrighetto commented:
Kind ping.

aeubanks commented:
precomitting the test would be nice so we can see the effect of this change

aeubanks commented:
also could you run this through https://llvm-compile-time-tracker.com/?

antoniofrighetto commented:
> precomitting the test would be nice so we can see the effect of this change

Tests are already in a separate commit (https://github.com/llvm/llvm-project/pull/101930/commits/2573b26d9dc2dcb53fe0492eccecd2f6f8486295).

> also could you run this through https://llvm-compile-time-tracker.com/?

Compile-time looks OK: https://llvm-compile-time-tracker.com/compare.php?from=2573b26d9dc2dcb53fe0492eccecd2f6f8486295&to=fd47d7f80291f5384c2ace760463cd39791f81ff&stat=instructions:u.

dtcxzyw commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1854,12 +1855,53 @@ bool MemCpyOptPass::processMemCpy(MemCpyInst *M, BasicBlock::iterator &BBI) {
+  if (isModSet(AA->getModRefInfo(M, MemoryLocation::getForSource(M)))) {
+    // On the off-chance the memmove clobbers src with previously memset'd
+    // bytes, the memmove may be redundant.
+    if (isMemMoveMemSetDependency(M)) {
```
```suggestion
    if (!M->isVolatile() && isMemMoveMemSetDependency(M)) {
```
Should bail out volatile memmoves.

dtcxzyw commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1854,12 +1855,53 @@ bool MemCpyOptPass::processMemCpy(MemCpyInst *M, BasicBlock::iterator &BBI) {
+  if (auto *Def = dyn_cast<MemoryDef>(SourceClobber))
+    MS = dyn_cast_or_null<MemSetInst>(Def->getMemoryInst());
+
+  if (!MS || MS->getParent() != M->getParent())
```
Please add a negative test for `MS->getParent() != M->getParent()`. Why we cannot do this if the memset and memmove are not in the same block?

antoniofrighetto replied:
I had limited this to the same block mostly as to avoid unnecessaries queries to the cached AA results, but I think it should be fine as long as there are no clobbering writes in between. Maybe worth understanding if this is the correct way to pursue, before dropping it.

nikic commented on llvm/test/Transforms/MemCpyOpt/memset-memmove-redundant-memmove.ll:
```diff
@@ -0,0 +1,88 @@
+; CHECK-NEXT:    ret i32 [[VAL]]
+;
+  %array = alloca [26 x i32], align 16
+  call void @llvm.memset.p0.i64(ptr noundef nonnull align 16 %array, i8 0, i64 104, i1 false)
```
Please remove irrelevant attributes like noundef nonnull.

nikic commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1853,12 +1854,54 @@ bool MemCpyOptPass::processMemCpy(MemCpyInst *M, BasicBlock::iterator &BBI) {
+    return false;
+
+  // The destination buffer must have been memset'd.
+  if (!BAA.isMustAlias(MS->getDest(), M->getDest()))
```
You also need to make sure that the source location is a subset of the memset memory. Currently, you just find a *potentially* clobbering memset, and then assume that it is fully clobbering. I think this is why https://github.com/dtcxzyw/llvm-opt-benchmark/pull/1135/files#diff-532ba66b66620dd644c14e91d18a3293194f3bfadfef06f698e87588e3d060ed is miscompiled.

nikic commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1853,12 +1854,54 @@ bool MemCpyOptPass::processMemCpy(MemCpyInst *M, BasicBlock::iterator &BBI) {
+
+  // No clobbering writes in between.
+  if (writtenBetween(MSSA, BAA, SourceLoc, MSSA->getMemoryAccess(MS),
+                     MemMoveAccess))
```
This check doesn't make sense to me. You have already checked above that SourceLoc is not clobbered using getClobberingMemoryAccess. What you need to verify now is that the destination location is not clobbered either. In your tests that would be `store i32 1, ptr %array` instead of `store i32 1, ptr %array.idx`.

It would probably be cleanest if you a) first checked that the memmove is of the form memmove(x, x+A, B) and then checked the location x with LocationSize A+B to cover both at the same time. Then you don't need writtenBefore and can handle the cross-BB case for free.

antoniofrighetto replied:
I see, indeed the first dominating clobbering access for the destination location should be the memset too. Could you please check that now this is more aligned with what you thought? Though, shouldn't the LocationSize be just B to check the source region (and not A+B)?

antoniofrighetto commented:
Does this look on track?

nikic commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1843,12 +1844,77 @@ bool MemCpyOptPass::processMemCpy(MemCpyInst *M, BasicBlock::iterator &BBI) {
+/// be removed.
+bool MemCpyOptPass::isMemMoveMemSetDependency(MemMoveInst *M) {
+  const auto &DL = M->getDataLayout();
+  MemSetInst *MS = nullptr;
```
Move MS declaration down.

nikic commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1843,12 +1844,77 @@ bool MemCpyOptPass::processMemCpy(MemCpyInst *M, BasicBlock::iterator &BBI) {
+  if (!BAA.isMustAlias(MS->getDest(), M->getDest()))
+    return false;
+
+  // The memmove is of form memmove(x, x + A, B).
```
It probably makes sense to check this first, before we do MSSA queries?

nikic commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1843,12 +1844,77 @@ bool MemCpyOptPass::processMemCpy(MemCpyInst *M, BasicBlock::iterator &BBI) {
+  if (!Source)
+    return false;
+  APInt Offset(DL.getIndexTypeSizeInBits(Source->getType()), 0);
+  auto MemMoveSize = MemoryLocation::getForSource(M).Size;
```
Already have SourceLoc above.

nikic commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1843,12 +1844,77 @@ bool MemCpyOptPass::processMemCpy(MemCpyInst *M, BasicBlock::iterator &BBI) {
+  // The first dominating clobbering MemoryAccess for the destination location
+  // needs to be the memset as well.
+  MemoryAccess *DestClobber =
+      MSSA->getWalker()->getClobberingMemoryAccess(FirstDef, DestLoc, BAA);
```
Can we do only a single MSSA query with the combined location instead of two?

antoniofrighetto replied:
Updated, thanks!

nikic commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1843,12 +1844,68 @@ bool MemCpyOptPass::processMemCpy(MemCpyInst *M, BasicBlock::iterator &BBI) {
+  if (!Source)
+    return false;
+  APInt Offset(DL.getIndexTypeSizeInBits(Source->getType()), 0);
+  auto MemMoveSize = SourceLoc.Size;
```
```suggestion
  LocationSize MemMoveSize = SourceLoc.Size;
```

nikic commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1843,12 +1844,68 @@ bool MemCpyOptPass::processMemCpy(MemCpyInst *M, BasicBlock::iterator &BBI) {
+  APInt Offset(DL.getIndexTypeSizeInBits(Source->getType()), 0);
+  auto MemMoveSize = SourceLoc.Size;
+  if (!Source->accumulateConstantOffset(DL, Offset) || Offset.isNegative() ||
+      Source->getPointerOperand() != M->getDest() || !MemMoveSize.hasValue())
```
Can move the getPointerOperand() check first, it's cheaper than accumulateConstantOffset...

antoniofrighetto replied:
Should have done earlier.

nikic commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1843,12 +1844,68 @@ bool MemCpyOptPass::processMemCpy(MemCpyInst *M, BasicBlock::iterator &BBI) {
+
+  if (!isModOrRefSet(BAA.getModRefInfo(MS, CombinedSourceLoc)) ||
+      !isModOrRefSet(BAA.getModRefInfo(MS, CombinedDestLoc)))
+    return false;
```
I don't think these getModRefInfo() queries are needed? This is implied by the isMustAlias call already.

A check I am missing though is that the memset size is sufficiently large. I think currently your transform would trigger on a partial memset?

antoniofrighetto replied:
Right, that one was missing, added a test for it too, thanks. While at it, I also added the case (and test for it) where x + A appears as a constantexpr (not sure if it's a case really needed though).

nikic commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1841,12 +1842,87 @@ bool MemCpyOptPass::processMemCpy(MemCpyInst *M, BasicBlock::iterator &BBI) {
+  // The memmove is of form memmove(x, x + A, B).
+  MemoryLocation SourceLoc = MemoryLocation::getForSource(M);
+  auto *MemMoveSourceOp = M->getSource();
+  auto *Source = dyn_cast<GetElementPtrInst>(MemMoveSourceOp);
```
```suggestion
  auto *Source = dyn_cast<GEPOperator>(MemMoveSourceOp);
```
handles both instruction and constexpr, no need to do both.

antoniofrighetto replied:
Thanks, updated.

nikic commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1841,12 +1842,74 @@ bool MemCpyOptPass::processMemCpy(MemCpyInst *M, BasicBlock::iterator &BBI) {
+  APInt Offset(DL.getIndexTypeSizeInBits(Source->getType()), 0);
+  LocationSize MemMoveLocSize = SourceLoc.Size;
+  if (Source->getPointerOperand() != M->getDest() ||
+      !MemMoveLocSize.hasValue() || Offset.isNegative() ||
```
Offset.isNegative() needs to be checked after accumulateConstantOffset.

antoniofrighetto replied:
Ugh, overlooked it before.

nikic commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1841,12 +1842,74 @@ bool MemCpyOptPass::processMemCpy(MemCpyInst *M, BasicBlock::iterator &BBI) {
+    return false;
+  }
+
+  const uint64_t MemMoveSize = MemMoveLocSize.getValue();
```
```suggestion
  uint64_t MemMoveSize = MemMoveLocSize.getValue();
```

nikic commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1841,12 +1842,74 @@ bool MemCpyOptPass::processMemCpy(MemCpyInst *M, BasicBlock::iterator &BBI) {
+  BatchAAResults BAA(*AA);
+  MemSetInst *MS = nullptr;
+  MemoryAccess *FirstDef = MemMoveAccess->getDefiningAccess();
+  MemoryAccess *DestClobber = MSSA->getWalker()->getClobberingMemoryAccess(
+      FirstDef, CombinedDestLoc, BAA);
```
```suggestion
  auto *DestClobber = dyn_cast<MemoryDef>(MSSA->getWalker()->getClobberingMemoryAccess(
      FirstDef, CombinedDestLoc, BAA));
  if (!DestClobber)
    return nullptr;
```
And then you can directly assign `MS` after that.

nikic commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1841,12 +1842,74 @@ bool MemCpyOptPass::processMemCpy(MemCpyInst *M, BasicBlock::iterator &BBI) {
+    return false;
+
+  // Memset length must be sufficiently large.
+  if (cast<ConstantInt>(MS->getLength())->getZExtValue() < MemMoveSize)
```
Can you please add a test where the memset size is not constant? We should use dyn_cast here.

antoniofrighetto replied:
Added, and used dyn_cast.

nikic commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1841,12 +1842,77 @@ bool MemCpyOptPass::processMemCpy(MemCpyInst *M, BasicBlock::iterator &BBI) {
+  uint64_t MemMoveSize = MemMoveLocSize.getValue();
+  LocationSize TotalSize =
+      LocationSize::precise(Offset.getZExtValue() + MemMoveSize);
+  MemoryLocation CombinedSourceLoc(MemMoveSourceOp, TotalSize);
```
Unused variable.

nikic commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1841,12 +1842,77 @@ bool MemCpyOptPass::processMemCpy(MemCpyInst *M, BasicBlock::iterator &BBI) {
+  LocationSize TotalSize =
+      LocationSize::precise(Offset.getZExtValue() + MemMoveSize);
+  MemoryLocation CombinedSourceLoc(MemMoveSourceOp, TotalSize);
+  MemoryLocation CombinedDestLoc(M->getDest(), TotalSize);
```
```suggestion
  MemoryLocation CombinedLoc(M->getDest(), TotalSize);
```
It's now both the Dest and Source Loc :)

nikic commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1841,12 +1842,77 @@ bool MemCpyOptPass::processMemCpy(MemCpyInst *M, BasicBlock::iterator &BBI) {
+    return false;
+
+  // The destination buffer must have been memset'd.
+  if (!BAA.isMustAlias(MS->getDest(), M->getDest()))
```
Can you please add a negative test where this check fails?

nikic reviewed (approved):
LGTM

antoniofrighetto committed:
[MemCpyOpt] Introduce test for PR101930 (NFC)

antoniofrighetto committed:
[MemCpyOpt] Drop dead `memmove` calls on `memset`'d source data

When a memmove happens to clobber source data, and such data have
been previously memset'd, the memmove may be redundant.
